Migrate to AVKit

j-maas opened the pull request:
`TODO`s mark places where I'm unsure about the migration.

The major todos are:
- [x] Reintroduce file type checking.
- [ ] Reintroduce exporting.
- [ ] Write temp video async.
- [ ] Fix frame generation on mouse drag. (I don't know precisely what the frame is for.)

The first commit to compile again is b1933cfa8b33740914cb1b598bfb7403b628a394.

afouse committed:
Code modernization to enable compiling in 64 bit. Elimination of File Uploading capability to get rid of ConnectionKit dependency. This means that log files and video feedback won't be possible anymore, but these capabilities aren't really used anymore. Updated BGHUDAppKit and Sparkle frameworks. Audio extraction is currently disabled.

afouse committed:
Remove Feedback item from help menu. Fix bug with applying filter to AFMovieView.

afouse committed:
Fix "Disable Video" button formatting.

afouse committed:
Update README

afouse committed:
Updating version number and .xib files

afouse committed:
Moving from QTKit to AVFoundation for Annotations

afouse committed:
Moving from QTKit to AVFoundation for AnnotationDocument

afouse committed:
Initial work on VideoProperties

afouse committed:
Finish moving VideoProperties to AVAsset

afouse committed:
Begin work on updating AppController

afouse committed:
Continued work on AppController. Expansion of VideoProperties to contain AVPlayer and AVPlayerItem instances.

afouse committed:
Updates to TimelineView.

afouse committed:
Updating DataSource. Creating NSCoder category to support loading in QTTime and QTTimeRange values.

afouse committed:
Updates to TimeCodedData and TimeCodedDataPoint

afouse committed:
Initial work on TimeSeriesData

afouse committed:
Update to TimeSeriesData for AVKit

j-maas committed:
Migrate many classes

j-maas committed:
More migrations (also using regex)

j-maas committed:
Migrate more

afouse committed:
Elimination of digital paper, interaction logs, and remote control

j-maas committed:
More migrations

j-maas committed:
Remove and stub out BGHUDAppKit

afouse committed:
Update MainMenu.xib to remove QTKit references

afouse committed:
Fixes for AVKit methods during loading

j-maas committed:
Convert from seconds with CMMakeTimeWithSeconds(seconds, 600)

Previously used CMMakeTime(seconds, 1000000).

j-maas committed:
Remove TODO for checking rounding method

Since we use the default method and it was unspecified for the old API, this should be fine.

afouse committed:
Temporary fix for default file creation async issue

afouse committed:
Additional avkit updates

afouse committed:
Fix for AVPlayerLayer retain/release

j-maas committed:
Remove QTMovieModernizer

j-maas committed:
Fix getting natural size of video

Previously, we tried to call tracksWithMedia on the AVPlayer, but it needs to be called on its AVAsset.

j-maas commented on Classes/Core/AppController.m:
```diff
@@ -3496,9 +3371,9 @@ - (void)updateDisplay:(NSTimer *)aTimer
-            QTTime normalizedStart = QTMakeTimeScaled([selectedAnnotation startTime], currentTime.timeScale);
-			if(QTTimeCompare(currentTime,[selectedAnnotation endTime]) == NSOrderedDescending)
+            CMTime currentTime = [mMovie currentTime];
+            CMTime normalizedStart = CMTimeConvertScale([selectedAnnotation startTime], currentTime.timescale, kCMTimeRoundingMethod_Default); // TODO: Check if QTMakeTimeScaled is correctly replacesd by CMTimeConvertScale
```
Is this replacement correct? I can't find any documentation on `QTMakeTimeScaled` to ensure that it did the same as `CMTimeConvertScale` does.

j-maas replied:
Note that there is the [same todo in `TranscriptView.m`](https://github.com/afouse/chronoviz/blob/avkit-attempt/Classes/Views/TranscriptView.m#L114).

afouse replied:
@Y0hy0h I looked into this a bit, and this is the correct replacement.

j-maas committed:
Remove unused variables

j-maas committed:
Fix broken frame generation method

j-maas committed:
Modernize deprecated version detection

j-maas committed:
Remove tensix frame generation case that appears to have no effect

The frame is generated but never assigned.

j-maas committed:
Remove unneeded frameDicts

j-maas committed:
Partly fix frame-by-frame stepping

Works on the default, blank video in a new project, but does not work after loading an actual video.

j-maas committed:
Replace QTMovie consistently with AVPlayer

j-maas committed:
Extract function from initFromFile for creating documentProperties

j-maas committed:
Initialize default documentProperties

j-maas committed:
Break out videoProperties init

j-maas committed:
Replace more QTMovie AVPlayer

j-maas committed:
Remove deleted file from project

j-maas committed:
Merge branch 'avkit-attempt'

j-maas committed:
Load annotation keyframe via VideoFrameLoader

j-maas committed:
Remove framework paths pointing outside the folder

j-maas committed:
Remove deprecated option 'ALWAYS_SEARCH_USER_PATHS' 

j-maas committed:
Assert that asset contains URL

j-maas committed:
Fix scrubbing precision

Previously videos would not scrub to the exact playhead position.

j-maas committed:
Nit: Use timeline range's start timescale

j-maas committed:
Remove waveform generation menu entry

j-maas committed:
Remove AudioExtractor, AudioVisualizer, and audio timelines

j-maas committed:
Close quick entry window when clicking outside

j-maas committed:
Work around crash on timeline keyframe generation

The AVPlayer was somehow not yet ready and caused the duration to be indefinite, leading to a NaN for `interval`.

j-maas committed:
Work around annotation document being nil when loading annotation

j-maas committed:
Work around video not being ready when updating timeline's playhead

j-maas committed:
Fix category text color in quick entry

Previously was black text on dark background. Is now white text.

j-maas committed:
Remove todo for AFHUDOutlineView

j-maas committed:
Clarify log todo

j-maas committed:
Fix todos with frame-by-frame stepping

j-maas committed:
Remove unused method

j-maas committed:
Remove todo for AFMovieView

There are no issues playing the video, so I assume that it is the correct replacement.

j-maas committed:
Reimplement looping

afouse committed:
Fix for annotation creation in plugins

afouse committed:
Fix for category filter window closing

afouse committed:
Add selection of annotation category as option for input to analysis plugins

j-maas committed:
Fix CMTime calculations for zooming in

j-maas committed:
Fix race condition for time display after clicking annotation

Previously, after seeking to the annotation's start time, the time display was updated to the current time too early when the seek was not yet completed. Now we use a completion handler to update the time after the seek was successful.

j-maas committed:
Simplify playback mode detection

j-maas committed:
Refactor frame generation method

j-maas committed:
Refactor seeks to set tolerance and use CMTIME_COMPARE_INLINE

j-maas committed:
Replace all uses of CMTimeCompare with CMTIME_COMPARE_INLINE via regex

j-maas committed:
Fix playhead dragging preview precision

Previously, dragging the playhead would only update the current time to a precision of a full second. It was not possible to see the frames in-between by dragging.

This was due to a timescale of 1 which leads to excessive rounding. I don't know where that timescale comes from.

afouse committed:
VideoDataSource file validation

afouse committed:
Update MovieViewer nib to fix AFMovieView link

j-maas commented:
It looks like this is already merged into `avkit`.
